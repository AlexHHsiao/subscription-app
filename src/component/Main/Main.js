import React, {Component, createRef} from 'react';
import ModalComponent from "./../../common/component/Modal/Modal";
import Spinner from "./../../common/component/Spinner/Spinner";
import {getCurrent, updateCurrent, getPreview} from "../../common/services/ApiService";
import Header from "../Header/Header";
import Subscription from "../Subscription/Subscription";

class Main extends Component {
    constructor(props) {
        super(props);

        this.state = {
            showSpinner: true,
            showModal: false,
            subscriptionData: null,
            modalData: {
                type: 'Error'
            },
            previewPrice: null
        };
    }

    componentDidMount() {
        getCurrent().then(
            response => {
                this.setState({
                    showSpinner: false,
                    subscriptionData: response
                });
            }).catch(error => {
            this.setState({
                showSpinner: false,
                showModal: true,
                modalData: {
                    type: 'Error',
                    title: 'Something went wrong',
                    message: error.statusText
                }
            });
        });
    }

    onCloseModal = () => {
        this.setState({
            showModal: false
        })
    };

    getPreview = (plan, seats) => {
        this.setState({
            showSpinner: true
        });
        const body = {
            plan,
            seats
        };
        getPreview(body).then(
            response => {
                this.setState({
                    showSpinner: false,
                    previewPrice: response.cost
                });
            }).catch(error => {
            this.setState({
                showSpinner: false,
                showModal: true,
                modalData: {
                    type: 'Error',
                    title: 'Something went wrong',
                    message: error.statusText
                }
            });
        });
    };

    getUpdate = (plan, seats) => {
        this.setState({
            showSpinner: true
        });
        const body = {
            plan,
            seats
        };
        updateCurrent(body).then(
            response => {
                this.setState((prevState) => ({
                    showSpinner: false,
                    subscriptionData: response,
                    showModal: true,
                    previewPrice: null,
                    modalData: {
                        type: 'Success',
                        title: 'Update Success',
                        prevSubscription: {...prevState.subscriptionData},
                        newSubscription: response
                    }
                }));
            }).catch(error => {
            this.setState({
                showSpinner: false,
                showModal: true,
                modalData: {
                    type: 'Error',
                    title: 'Something went wrong',
                    message: error.statusText
                }
            });
        });
    };

    render() {
        const {showSpinner, showModal, modalData, subscriptionData, previewPrice} = this.state;

        return (
            <>
                <Spinner open={showSpinner}/>
                <ModalComponent open={showModal} onClose={this.onCloseModal} modalData={modalData}/>
                {subscriptionData ? (
                    <>
                        <Header getPreview={() => this.getPreview(subscriptionData.plan, subscriptionData.seats)}/>
                        <div className='container mt-5'>
                            <div className='row'>
                                <div className='col-12'>
                                    <Subscription
                                        subscriptionData={subscriptionData}
                                        getPreview={this.getPreview}
                                        getUpdate={this.getUpdate}
                                        previewPrice={previewPrice}
                                    />
                                </div>
                            </div>
                        </div>
                    </>
                ) : ''
                }
            </>
        );
    }
}

export default Main;
