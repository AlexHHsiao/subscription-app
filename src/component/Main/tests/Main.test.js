import React from 'react';
import {render} from '@testing-library/react';
import Main from '../Main';

test('<Main/>', () => {
    const {container: {firstChild}} = render(<Main/>);
    expect(firstChild).toMatchSnapshot();
});
