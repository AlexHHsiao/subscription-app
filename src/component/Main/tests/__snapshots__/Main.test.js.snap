// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main/> 1`] = `
<div
  class="spinnerContainer"
>
  <div
    class="spinnerContent"
  >
    <div
      class="spinner-border"
      style="width: 75px; height: 75px;"
    />
  </div>
</div>
`;
